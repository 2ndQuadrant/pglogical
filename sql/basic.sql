SET synchronous_commit = on;

-- Schema setup

CREATE TABLE demo (
	seq serial primary key,
	tx text,
	ts timestamp,
	jsb jsonb,
	js json,
	ba bytea
);

SELECT 'init' FROM pg_create_logical_replication_slot('regression_slot', 'pglogical_output');

-- Queue up some work to decode with a variety of types

INSERT INTO demo(tx) VALUES ('textval');
INSERT INTO demo(ba) VALUES (BYTEA '\xDEADBEEF0001');
INSERT INTO demo(ts, tx) VALUES (TIMESTAMP '2045-09-12 12:34:56.00', 'blah');
INSERT INTO demo(js, jsb) VALUES ('{"key":"value"}', '{"key":"value"}');

-- Simple decode with text-format tuples
--
-- It's still the logical decoding binary protocol and as such it has
-- embedded timestamps, and pglogical its self has embedded LSNs, xids,
-- etc. So all we can really do is say "yup, we got the expected number
-- of messages".
SELECT count(data) FROM pg_logical_slot_peek_binary_changes('regression_slot',
	NULL, NULL,
	'expected_encoding', 'UTF8',
	'min_proto_version', '1',
	'max_proto_version', '1',
	'startup_params_format', '1');

-- ... and send/recv binary format
-- The main difference visible is that the bytea fields aren't encoded
SELECT count(data) FROM pg_logical_slot_peek_binary_changes('regression_slot',
	NULL, NULL,
	'expected_encoding', 'UTF8',
	'min_proto_version', '1',
	'max_proto_version', '1',
	'startup_params_format', '1',
	'binary.want_binary_basetypes', '1',
	'binary.basetypes_major_version', (current_setting('server_version_num')::integer / 100)::text);

SELECT 'drop' FROM pg_drop_replication_slot('regression_slot');

DROP TABLE demo;
