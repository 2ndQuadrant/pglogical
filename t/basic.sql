-- basic builtin datatypes
CREATE OR REPLACE FUNCTION public.pg_xlog_wait_remote_apply(i_pos pg_lsn, i_pid integer) RETURNS VOID
AS $FUNC$
BEGIN
    WHILE EXISTS(SELECT true FROM pg_stat_get_wal_senders() s WHERE s.replay_location < i_pos AND (i_pid = 0 OR s.pid = i_pid)) LOOP
                PERFORM pg_sleep(0.01);
        END LOOP;
END;$FUNC$ LANGUAGE plpgsql;

SELECT pglogical.replicate_ddl_command($$
	CREATE TABLE public.basic_dml (
		id serial primary key,
		other integer,
		data text,
		something interval
	);
$$);

SELECT * FROM pglogical.replication_set_add_table('default', 'basic_dml');

SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);

-- check basic insert replication
INSERT INTO basic_dml(other, data, something)
VALUES (5, 'foo', '1 minute'::interval),
       (4, 'bar', '12 weeks'::interval),
       (3, 'baz', '2 years 1 hour'::interval),
       (2, 'qux', '8 months 2 days'::interval),
       (1, NULL, NULL);
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);

SELECT id, other, data, something FROM basic_dml ORDER BY id;

