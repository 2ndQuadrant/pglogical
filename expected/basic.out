-- basic builtin datatypes
SELECT * FROM pglogical_regress_variables()
\gset
\c :provider_dsn
SELECT pglogical.replicate_ddl_command($$
	CREATE TABLE public.basic_dml (
		id serial primary key,
		other integer,
		data text,
		something interval
	);
$$);
 replicate_ddl_command 
-----------------------
 t
(1 row)

SELECT * FROM pglogical.replication_set_add_table('default', 'basic_dml');
 replication_set_add_table 
---------------------------
 t
(1 row)

SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
 pg_xlog_wait_remote_apply 
---------------------------
 
(1 row)

\c :subscriber_dsn
ALTER TABLE public.basic_dml ADD COLUMN subonly integer;
ALTER TABLE public.basic_dml ADD COLUMN subonly_def integer DEFAULT 99;
\c :provider_dsn
-- check basic insert replication
INSERT INTO basic_dml(other, data, something)
VALUES (5, 'foo', '1 minute'::interval),
       (4, 'bar', '12 weeks'::interval),
       (3, 'baz', '2 years 1 hour'::interval),
       (2, 'qux', '8 months 2 days'::interval),
       (1, NULL, NULL);
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
 pg_xlog_wait_remote_apply 
---------------------------
 
(1 row)

\c :subscriber_dsn
SELECT id, other, data, something FROM basic_dml ORDER BY id;
 id | other | data |    something     
----+-------+------+------------------
  1 |     5 | foo  | @ 1 min
  2 |     4 | bar  | @ 84 days
  3 |     3 | baz  | @ 2 years 1 hour
  4 |     2 | qux  | @ 8 mons 2 days
  5 |     1 |      | 
(5 rows)

-- update one row
\c :provider_dsn
UPDATE basic_dml SET other = '4', data = NULL, something = '3 days'::interval WHERE id = 4;
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
 pg_xlog_wait_remote_apply 
---------------------------
 
(1 row)

\c :subscriber_dsn
SELECT id, other, data, something FROM basic_dml ORDER BY id;
 id | other | data |    something     
----+-------+------+------------------
  1 |     5 | foo  | @ 1 min
  2 |     4 | bar  | @ 84 days
  3 |     3 | baz  | @ 2 years 1 hour
  4 |     4 |      | @ 3 days
  5 |     1 |      | 
(5 rows)

-- update multiple rows
\c :provider_dsn
UPDATE basic_dml SET other = id, data = data || id::text;
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
 pg_xlog_wait_remote_apply 
---------------------------
 
(1 row)

\c :subscriber_dsn
SELECT id, other, data, something FROM basic_dml ORDER BY id;
 id | other | data |    something     
----+-------+------+------------------
  1 |     1 | foo1 | @ 1 min
  2 |     2 | bar2 | @ 84 days
  3 |     3 | baz3 | @ 2 years 1 hour
  4 |     4 |      | @ 3 days
  5 |     5 |      | 
(5 rows)

\c :provider_dsn
UPDATE basic_dml SET other = id, something = something - '10 seconds'::interval WHERE id < 3;
UPDATE basic_dml SET other = id, something = something + '10 seconds'::interval WHERE id > 3;
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
 pg_xlog_wait_remote_apply 
---------------------------
 
(1 row)

\c :subscriber_dsn
SELECT id, other, data, something, subonly, subonly_def FROM basic_dml ORDER BY id;
 id | other | data |     something      | subonly | subonly_def 
----+-------+------+--------------------+---------+-------------
  1 |     1 | foo1 | @ 50 secs          |         |          99
  2 |     2 | bar2 | @ 84 days -10 secs |         |          99
  3 |     3 | baz3 | @ 2 years 1 hour   |         |          99
  4 |     4 |      | @ 3 days 10 secs   |         |          99
  5 |     5 |      |                    |         |          99
(5 rows)

-- delete one row
\c :provider_dsn
DELETE FROM basic_dml WHERE id = 2;
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
 pg_xlog_wait_remote_apply 
---------------------------
 
(1 row)

\c :subscriber_dsn
SELECT id, other, data, something FROM basic_dml ORDER BY id;
 id | other | data |    something     
----+-------+------+------------------
  1 |     1 | foo1 | @ 50 secs
  3 |     3 | baz3 | @ 2 years 1 hour
  4 |     4 |      | @ 3 days 10 secs
  5 |     5 |      | 
(4 rows)

-- delete multiple rows
\c :provider_dsn
DELETE FROM basic_dml WHERE id < 4;
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
 pg_xlog_wait_remote_apply 
---------------------------
 
(1 row)

\c :subscriber_dsn
SELECT id, other, data, something FROM basic_dml ORDER BY id;
 id | other | data |    something     
----+-------+------+------------------
  4 |     4 |      | @ 3 days 10 secs
  5 |     5 |      | 
(2 rows)

-- truncate
\c :provider_dsn
TRUNCATE basic_dml;
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
 pg_xlog_wait_remote_apply 
---------------------------
 
(1 row)

\c :subscriber_dsn
SELECT id, other, data, something FROM basic_dml ORDER BY id;
 id | other | data | something 
----+-------+------+-----------
(0 rows)

-- copy
\c :provider_dsn
\COPY basic_dml FROM STDIN WITH CSV
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
 pg_xlog_wait_remote_apply 
---------------------------
 
(1 row)

\c :subscriber_dsn
SELECT id, other, data, something FROM basic_dml ORDER BY id;
  id  | other | data | something 
------+-------+------+-----------
 9000 |     1 | aaa  | @ 1 hour
 9001 |     2 | bbb  | @ 2 years
 9002 |     3 | ccc  | @ 3 mins
 9003 |     4 | ddd  | @ 4 days
(4 rows)

\c :provider_dsn
\set VERBOSITY terse
SELECT pglogical.replicate_ddl_command($$
	DROP TABLE public.basic_dml CASCADE;
$$);
NOTICE:  drop cascades to table public.basic_dml membership in replication set default
 replicate_ddl_command 
-----------------------
 t
(1 row)

